storage/kafka/datasource.go

```diff
@@ -0,0 +1,381 @@
+			// Get execution configuration
+			hosts, ports, err := config.GetIPAddressList(dbConfig, "brokers", config.WithDefault([]interface{}{[]string{"localhost"}, []int{9092}}))
+			if err != nil {
+				return nil, errors.Wrap(err, "couldn't get address")
+			}
+			topic, err := config.GetString(dbConfig, "topic")
+			if err != nil {
+				return nil, errors.Wrap(err, "couldn't get topic")
+			}
+			startOffset, err := config.GetInt(dbConfig, "startOffset", config.WithDefault(-1))
+			if err != nil {
+				return nil, errors.Wrap(err, "couldn't get start offset")
+			}
+			batchSize, err := config.GetInt(dbConfig, "batchSize", config.WithDefault(1))
+			if err != nil {
+				return nil, errors.Wrap(err, "couldn't get batch size")
+			}
+			decodeAsJSON, err := config.GetBool(dbConfig, "json", config.WithDefault(false))
+			if err != nil {
+				return nil, errors.Wrap(err, "couldn't get json option")
+			}
+
+			brokers := make([]string, len(hosts))
+			for i := range hosts {
+				brokers[i] = fmt.Sprintf("%s:%d", hosts[i], ports[i])
+			}
+
+			return &DataSource{
+				brokers:      brokers,
+				topic:        topic,
+				partition:    partition,
+				startOffset:  startOffset,
+				batchSize:    batchSize,
+				decodeAsJSON: decodeAsJSON,
+				alias:        alias,
+				stateStorage: matCtx.Storage,
+			}, nil
+		},
+		nil,
+		availableFilters,
+		metadata.Unbounded,
+		partitions,
+	)
+}
+
+// NewDataSourceBuilderFactoryFromConfig creates a data source builder factory using the configuration.
+func NewDataSourceBuilderFactoryFromConfig(dbConfig map[string]interface{}) (physical.DataSourceBuilderFactory, error) {
+	partitions, err := config.GetInt(dbConfig, "partitions", config.WithDefault(1))
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't get primaryKeys")
+	}
+
+	return NewDataSourceBuilderFactory(partitions), nil
+}
+
+var offsetPrefix = []byte("$kafka_offset$")
+var tokenQueuePrefix = []byte("$token_queue$")
+
+const initialTokenCount = 5
+
+func (ds *DataSource) Get(ctx context.Context, variables octosql.Variables, streamID *execution.StreamID) (execution.RecordStream, *execution.ExecutionOutput, error) {
+	tx := storage.GetStateTransactionFromContext(ctx).WithPrefix(streamID.AsPrefix())
+
+	dialer := &kafka.Dialer{
+		Timeout:   10 * time.Second,
+		DualStack: true,
+	}
+
+	kafkaConfig := kafka.ReaderConfig{
+		Brokers:   ds.brokers,
+		Topic:     ds.topic,
+		Dialer:    dialer,
+		Partition: ds.partition,
+		MinBytes:  10e1,
+		MaxBytes:  10e7,
+	}
+
+	r := kafka.NewReader(kafkaConfig)
+
+	rs := &RecordStream{
+		stateStorage: ds.stateStorage,
+		streamID:     streamID,
+		kafkaReader:  r,
+		startOffset:  ds.startOffset,
+		batchSize:    ds.batchSize,
+		decodeAsJSON: ds.decodeAsJSON,
+		alias:        ds.alias,
+	}
+	err := rs.loadOffset(tx)
+	if err != nil {
+		return nil, execution.NewExecutionOutput(execution.NewZeroWatermarkGenerator()), errors.Wrapf(err, "couldn't load kafka partition %d offset", ds.partition)
```

Comment:
The concept in other datasources is return `nil` instead of `execution.NewExecutionOutput(execution.NewZeroWatermarkGenerator())` with early return due to error

Reply:
oh yeah, sorry, changing